README
FrameNet-WordNet-Detour

VERSION 0.99c

DESCRIPTION

The FrameNet-WordNet-Detour takes one or more synsets as input and
maps each of them to a set of weighted frames.

PREREQUISITS

CPAN-Modules:
- WordNet::QueryData (>= 1.37)
- WordNet::Similarity::path (>=0.10)
- XML::TreeBuilder (>=3.09)
- Getopt::Long (>= 2.34)
- Pod::Usage (>=1.00)
- File::Spec (>=0.86)
- Test::More

Additionally:
- WordNet (needed also for the WordNet::* modules)
- FrameNet (>= 1.1, we tested on 1.1 and 1.2)

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

With 

    perl Makefile.pl PREFIX=/some/directory

you can target the installation in any directory you want.

The testing code makes use of the two environment variables $FNHOME
and $WNHOME, which should be set to the FrameNet resp. WordNet
directory. The testing code will not work without these variables,
even though it is possible to use the module later on without them.

As a workaround, we propose to temporarly set the variables (or to
skip the tests, but that probably not a good advice ... )

COPYRIGHT AND LICENCE

Copyright (C) 2005 Aljoscha Burchardt, Nils Reiter

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for FrameNet-WordNet-Detour

1.0
     - Fixed a bug in the weighting algorithm
     - Introduced the method version().

0.99c 10/14/06 
     - Fixed a bug in the documentation: get_message and get_name
       did not work.

0.99b 10/11/06
     - VERY small (but important) bugfix ...

0.99 10/10/06
     - Implemented a couple of helper methods in the Frame and 
       Data class to get rid of the dependency on
       Class::MethodMaker. 
     - Testing stops if $WNHOME or $FNHOME is not set and gives
       a readable error message.
     - Fixed a bug in the limited condition. In the division for 
       making the results comparable in different runs, the factor
       was wrongly calculated in limited mode.

0.98 02/25/06
     - Switched to the named parameter style
       (-wnhome => "..."). This means, that old scripts have to be
       updated. Additionally, one can give the whole QueryData object
       (-wnquerydata => $obj).
     - due to licensing problems, the LU cache is removed from the
       package. Again, one needs FrameNet to install the detour.
     - removed all the Growl code (well, it was mostly useless...).
       Along with that, we removed the customize-script to make it 
       more reliable and easy to maintain - we're approaching 1.0, you
       know ;-)


0.97 10/11/05
     - Using of File::Spec for platform independency
     - the command line script now has growl capabilities 
       if running on Mac OS X. Just install the perl module Mac::Growl
       and use the parameter "--growl". Find out more about Growl
       on http://www.growl.info/. The growl code is automatically
       included at make time if the module Mac::Growl is present       
     - Redesign of FrameNet::WordNet::Detour::Frame with usage 
       of Class::MethodMaker
     - Redesign of FrameNet::WordNet::Detour::Data with usage 
       of Class::MethodMaker
     - More documentation
     - the command line script is now able to print also the
       frame evoking element and its similarity
     - Cache is now on Version 0.91 (due to the Changes in
       Detour::Frame and Detour::Data)
     - detour is no more static script, but can be configured while
       installing (make) using the Customizer. It is a pattern that
       allows the customization of perl code with make. More
       information about this can be found on
       www.coli.uni-saarland.de/~reiter/index.php4?action=perl#custom. 
     - (a lot) more reliable and wide tests
     - the LU cache (based on FrameNet 1.2) is now part of the
       module (script/make-luhash.PL). Therefore, it can be installed 
       without FrameNet. The hash will be created with the
       installation

0.96 09/27/05
     - Fixed several bugs in the demo code
     - get_all_framenames in the Data class now works and returns a reference
     - Fixed a bug in Detour.pm that causes WordNet::QueryData to produce
       a lot of error messages	
     - Fixed a bug: 'word#pos' now returns a list reference with 
       "empty result" if no result found  
     - query() now recognizes immediately if no synset is given and returns a nice
       error message
     - updated documentation of the detour script

0.95 08/22/05
     - Fixed bug, which resulted in choosing the wrong cache
     - Better Accessibility of the interesting properties via the
       Data-object. Additional methods:
       * get_number_of_frames
       * get_weight 
       * get_fees
       * get_frame
       * get_best_weight
     - Better documentation of the Data-object
     - Fixed a bug in the command line script ("-l" did not run
       the detour in limited mode)
     - Fixed a bug in the verbose output
     - Command line interface is now called "detour" instead of "Detour"

0.94 08/02/05
     - Fixed Makefile bugs
     - It is now possible to give the path to WordNet and FrameNet 
       as arguments to new()

0.93 07/23/05
     - Improved Caching
     - Semantics of WNHOME is now identically with it in other 
       WordNet-related modules
     - Detour can now initialize its own QueryData and Similarity
       objects. In the new-method, they should left out by now. This
       improves the performance of your scripts severly, since queries
       that are already cached are answered now without need of the
       QueryData-object, which takes around 15 seconds to load

0.92 07/22/05
     - More documentation
     - Introduced a command-line interface called Detour:
       * located in bin/
       * shows the basic use of the Detour
       * requires additional modules: Getopt::Long and Pod::Usage
     - Improvements on the robustness of the Detour

0.91 07/22/05
     - Better testing capabilities
     - More compatibility with CPAN
     - More (and better) documentation

0.90 07/21/05
     - Initial Release



t/00-load.t.PL
use Config;
use File::Spec;

my $fname = shift;
my $hashfilename =  File::Spec->
  catfile((File::Spec->tmpdir),
	  (defined $ENV{'USER'}?
	   $ENV{'USER'}:
	   'user').
	  '-FrameNet-WordNet-Detour-0.91-luhash.dat');

open OUT, ">$fname" or die "Can't create 00-load.t: $!"; 
chmod(0755, "$fname"); 

print OUT $Config{startperl}, " -w \n"; 

print OUT q{

use Test::More tests => 5;
use File::Spec;

isnt($ENV{'WNHOME'}, "", "Testing if \$WNHOME is set");
isnt($ENV{'FNHOME'}, "", "Testing if \$FNHOME is set");
SKIP: {
  skip "\$WNHOME not set",1 unless $ENV{'WNHOME'} ne "";
  ok(-e $ENV{'WNHOME'}, "Testing if \$WNHOME exists");
}
SKIP: {
  skip "\$FNHOME not set",1 unless $ENV{'FNHOME'} ne "";
  ok(-e $ENV{'FNHOME'}, "Testing if \$FNHOME exists");
}

};

print OUT "\nok(-e \"$hashfilename\", \"Testing if LU hash was installed correctly\")\n";


script/customize.PL
use Config;

# customize script
# Copyright 2005 Nils Reiter. All rights reserved.
#
# This script is free software; 
# you can redistribute it and/or modify it under 
# the same terms as Perl itself.
#
# see the file README for informations how to use this

my $in = shift;
my $out = shift;

customize($out, $in);

sub customize {
  my $outfile = shift;
  my $protofile = shift;
  my $props = {};

  open IN, $protofile;
  open OUT, ">$outfile" or die "Can't create $outfile: $!"; 
  print OUT $Config{startperl}, " -w \n";  
  
  while(<IN>) {
    # Found customization information that is used in the following
    if (s/#CUSTOM_BEGIN#(\w+)#([^#]+)#CUSTOM_END#//) {
      #print STDERR "Property $1 registered\n";
      $props->{$1}->{code} = $2;
      my $p = $1;
      $props->{$p}->{value} = eval $props->{$p}->{code};
      $props->{$p}->{re_line} = "#".$p."_BEGIN#.*#".$p."_END#";
      $props->{$p}->{re_single} = "#".$p."_BEGIN#|#".$p."_END#";
    }
    # Each line is searched for each propery
    # Not tested with a lot of properties
    foreach my $p (keys %$props) {
      if (/$props->{$p}->{re_line}/) {
	s/$props->{$p}->{re_single}//g if ($props->{$p}->{value});
	s/$props->{$p}->{re_line}\n//g if (! $props->{$p}->{value}); 
      }
    };
    print OUT $_;
  }
  close IN;
  close OUT;
};


lib/FrameNet/WordNet/Detour/Data.pm
package FrameNet::WordNet::Detour::Data;

require Exporter;
our @ISA = qw(Exporter);
our $VERSION = "0.99";

use strict;
use warnings;
use FrameNet::WordNet::Detour::Frame;

sub new {
  my $class = shift;
  my $this = {};
  
  bless($this, $class);
  
  $this->{'s2f-result'} = shift;
  $this->{'raw'} = $this->{'s2f-result'}{'raw'};
  $this->{'sorted'} = $this->{'s2f-result'}{'sorted'};
  my @warr = reverse (sort { $a <=> $b } keys %{$this->{'sorted'}});
  $this->{'weights'} = \@warr;
  $this->query(shift);
  $this->message(shift) || $this->message('OK');

  return $this;
}

sub query {
    my ($self,$query) = @_;
    if (defined $query) {
	$self->{'query'} = $query;
    };
    return $self->{'query'};
}

sub get_query {
    my ($self,$query) = @_;
    if (defined $query) {
	return $self->query($query);
    };
    return $self->query;
}

sub message {
    my ($self,$msg) = @_;
    if (defined $msg) {
	$self->{'message'} = $msg;
    };
    return $self->{'message'};
}

sub get_message {
    my ($self,$msg) = @_;
    if (defined $msg) {
	$self->message($msg);
    };
    return $self->message;
}

# Checks, wether the query returned some useful results or not. 
# If not, one should check the error message via &get_message. 
# Could be better, by checking deeper in the data structure,
# if there is a 'Frame'-Object.
sub isOK {
  my $self = shift;
  return 1 if ($self->{'message'} eq 'OK');
  return 0;
};

sub is_ok {
    my $self = shift;
    return $self->isOK;
}

sub get_fees {
  my $self = shift;
  my $_frame = shift;
  my $f = $self->get_frame($_frame);
  return $f if ($f == -1);
  return $f->fees;
};

sub get_weight {
  my $self = shift;
  my $frame = shift;
  my $f = $self->get_frame($frame);

  return $f if ($f == -1);
  return $f->weight;
};

sub get_weights {
  my $self = shift;
  return $self->{'weights'};
};

# not working
sub get_delta {
  my $self = shift;
  my $frame = shift;
  

  my $w0 = $self->{'raw'}->{$frame}->{'weight'};
  my $w1 = 0;
  my @weights = ( sort { $a <=> $b } (keys %{$self->{'sorted'}}));



  for(my $i = 0; $i < scalar @weights; $i++) {
    $w1 = $weights[$i+1] if ($w0 == $weights[$i] && 
			     exists($weights[$i+1]));
  };
  return int((($w0 - $w1)*1000)+0.5)/1000;
};

sub get_number_of_frames {
  my $self = shift;
  return scalar (keys %{$self->{'raw'}});
};

# Returns a reference to an array containing the arg1 
# best frames (as Frame-Objects). Frames are sorted according
# to their weight.
# If arg1 is not given, the best frame will be returned.
# Works always on the first (e.g. 0th) synset.
sub get_best_frames {
  my $self = shift;
  my $n = 0;
  my $m = shift || 1;

  my $ResultsByWeight = $self->{'sorted'};

  my $ResultList = [];

  my $result_counter = 1;
  
  foreach my $weight (reverse(sort(keys %$ResultsByWeight))) {
    if ($result_counter <= $m) {
      foreach my $frame (keys %{$ResultsByWeight->{$weight}}) {
	push (@$ResultList, $ResultsByWeight->{$weight}->{$frame});
      };
    };
    $result_counter++;
  };
  return $ResultList;
};

# WORKS
sub get_best_framenames {
  my $self = shift;
  my $m = shift || 1;
  my $frames = $self->get_best_frames($m);
  my @arr =  map($_->name, @$frames);
  return \@arr;
};

# Returns a list of all found frames.
sub get_all_framenames ($) {
  my $self = shift;
  my $tmp = {};
  foreach my $frame (keys %{$self->{'raw'}}) {
      $tmp->{$frame} = 1;
  }
  my @ret = keys %$tmp; 
  return \@ret;
};

sub get_all_frames {
  my $self = shift;
  my $ResultsByWeight = $self->{'sorted'};
  my $ResultList = [];

  foreach my $weight (reverse(sort(keys %$ResultsByWeight))) {
    foreach my $frame (keys %{$ResultsByWeight->{$weight}}) {
      push (@$ResultList, $ResultsByWeight->{$weight}->{$frame});
    };
  };
  return $ResultList;
};

sub get_frame {
  my $self = shift;
  my $frame = shift;
  #print STDERR $frame."!!!";
  return $self->{'raw'}->{$frame} if (exists($self->{'raw'}->{$frame}));
  return $self->{'raw'}->{lc($frame)} if (exists($self->{'raw'}->{lc($frame)}));
  return $self->{'raw'}->{ucfirst($frame)} if (exists($self->{'raw'}->{ucfirst($frame)}));
  return -1;
};

sub get_best_weight {
  my $self = shift;
  my $w = $self->get_weights;
  return $w->[0];
};

sub get_frames_with_weight {
  my $self = shift;
  my $weight = shift;
  my @l = keys %{$self->{'sorted'}->{$weight}};
  return \@l;
};

__END__

## DOCUMENTATION ##

=head1 NAME

FrameNet::WordNet::Detour::Data - A class representing the results of the Detour.

=head1 SYNOPSIS

  use FrameNet::WordNet::Detour;

  my $wn = WordNet::QueryData->new($WNSEARCHDIR);
  my $sim = WordNet::Similarity::path->new ($wn);
  my $detour = FrameNet::WordNet::Detour->new($wn,$sim);

  my $result = $detour->query($synset);

  $result->is_ok;   # Returns whether there were problems in the run
  $result->message; # Returns 'Ok' or an error message

  # All frames are returned as lists of 
  # L<FrameNet::WordNet::Detour::Frame> objects


  $result->get_best_frames; # Returns the frames 
                            # with the highest weight
  $result->get_best_frames(3); # Returns the frames
                               # with the three highest weights
  $result->get_all_frames; # Returns all resulting frames


  $result->get_best_framenames;    # Returns the names
                                   # of the highest weighted frames
  $result->get_best_framenames(3); # Returns the names of the frames
                                   # with the three highest weights
  $result->get_all_framenames;     # Returns the names of all frames.

=head1 METHODS

=over 4

=item get_frame FRAME

Returns the frame $string as a FrameNet::WordNet::Detour::Frame-object. Returns -1 if FRAME is not in the result (Pay attention on lower/upper case, we look a bit around, but that could be a hard-to-find error).

=item get_best_frames [ NUMBER ] 

Returns the frames with the highest weight (as Frame-objects). Optional: If you specify $number, the method returns the $number highest rated frames.

=item get_all_frames

Returns all found frames as Frame-objects.

=item query

Returns the query-synset as string.

=item message

Returns an eventual error message.

=item get_fees FRAME

Returns the frame evoking elements for the given frame. Returns -1 if FRAME was not in the result (Pay attention on lower/upper case, we look a bit around, but that could be a hard-to-find error).

=item get_weight FRAME

Returns the weight of the given frame. Returns -1 if FRAME was not in the result(Pay attention on lower/upper case, we look a bit around, but that could be a hard-to-find error).

=item get_weights

Returns a reference to an array of all weights that appeared in this run. The weights are sorted in descending order (normally, one is interested in the best e.g. highest values instead of the bad ones).

=item get_number_of_frames

Returns the overall number of frames found for this specific query:

  # The first variant returns all frames, because even in the worst 
  # case (that each frame has a different weight) we get all weight
  # classes. It does not matter if one specifies a number greater 
  # that the numer of existing frames, 
  # so the second variant leads normally to the same results.

  $result->get_best_frames($result->get_number_of_frames) 

  $result->get_best_frames(1000);

=item get_frames_with_weight WEIGHT

Returns a reference to a list of the frames with the given weight. You should notice, that you have to give the exact weight - e.g. like in the get_weights-array. Rounded values will not find anything.

=back

=head1 BUGS

Please report bugs to L<mailto:reiter@cpan.org>.

=head1 COPYRIGHT

Copyright 2005 Aljoscha Burchardt and Nils Reiter. All rights reserved.

This module is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut


lib/FrameNet/WordNet/Detour/Frame.pm
package FrameNet::WordNet::Detour::Frame;

require Exporter;
our @ISA = qw(Exporter);
our $VERSION = "0.99";

use strict;
use warnings;

sub new {
    my $class = shift;
    my $self = { 'fees' => [],
		 'sims' => [] };
    return bless $self, $class;
}

sub fees {
    my $self = shift;
    return $self->{'fees'};
}

sub sims {
    my $self = shift;
    return $self->{'sims'};
}

sub fees_add {
    my ($self, $fee) = @_;
    push(@{$self->{'fees'}},$fee);
}

sub sims_add {
    my ($self, $sim) = @_;
    push(@{$self->{'sims'}},$sim);
}

sub add_weight {
  my $self = shift;
  my $w = shift;
  $self->{'weight'} += $w;
}

sub weight {
    my ($self,$w) = @_;
    if (defined $w) {
	$self->{'weight'} = $w;
    };
    return $self->{'weight'};
};

sub name {
    my ($self,$name) = @_;
    if (defined $name) {
	$self->{'name'} = $name;
    };
    return $self->{'name'};
};

sub get_name {
    my ($self,$name) = @_;
    if (defined $name) {
	return $self->name($name);
    };
    return $self->name;
};

sub get_weight {
    my ($self,$weight) = @_;
    if (defined $weight) {
	return $self->weight($weight);
    };
    return $self->weight;
};



1;


__END__


## DOCUMENTATION ##

=head1 NAME

FrameNet::WordNet::Detour::Frame - A class representing one single frame.

=head1 SYNOPSIS

  my $frame = {$result->get_best_frames}->[0];

  print "Frame ".$frame->name."\n";
  print "Weight ".$frame->weight."\n";
  print "Fees: ".join(",", @{$frame->fees});

=head1 METHODS

Note: Some of the methods allow writing of properties. This is only used during the creation of the module and should not be used and needed at all.

=over

=item new

=item name

If an argument is provided, the name of the frame is set to that value.  The method returns the name (after assignment to a provided value, if appropriate).

=item weight

Same as C<name>, except that it works for the weight instead of the name.

=item fees

Several Uses: If called without arguments, it returns the list of frame evoking elements in list context and a reference to that list in scalar context. Can also be called with an argument, that list B<replaces> the current value.

=item sims

Same as C<fees>, except that it works on the list of similarities of the fees.

=back

=head1 BUGS

Please report bugs to L<mailto:reiter@cpan.org>.

=head1 COPYRIGHT

Copyright 2005 Aljoscha Burchardt and Nils Reiter. All rights reserved.

This module is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;
use File::Spec;

my $VCACHE = 0.91;

WriteMakefile(
	      NAME                => 'FrameNet::WordNet::Detour',
	      AUTHOR              => 'Nils Reiter <reiter@cpan.org>',
	      VERSION_FROM        => 'lib/FrameNet/WordNet/Detour.pm',
	      ABSTRACT_FROM       => 'lib/FrameNet/WordNet/Detour.pm',
	      PREREQ_PM => 
	      { 'Test::More' => 0,
		'WordNet::QueryData' => 1.37,
		'WordNet::Similarity::path' => 0.10,
		'XML::TreeBuilder' => 3.09,
		'Getopt::Long' => 2.34,
		'Pod::Usage' => 1.00,
		'File::Spec' => 0.86 },

	      EXE_FILES => 
	      [ 'script/detour' ],

	      dist => 
	      { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },

	      clean => 
	      { FILES => 'FrameNet-WordNet-Detour-*' },

	      realclean =>
	      { FILES => '*~ blib' }
	     );
